### PR TITLE
Add KASAN support

### PR DESCRIPTION
**Pull request by tavip**
This PR is a port of @rodionov's KASAN patch set from the Android kernel. It adds generic KASAN support to LKL and enables it for x86_64 only for now. It also adds a CI test for it that basically runs the KASAN kunit tests.

### CODE REVIEW

**Comment by tavip** [comment]
@thehajime @rodionov the PR is ready for review.

Not sure if there is a good fix for the checkpatch failures, any ideas are welcome. The issue is that loglvl is now passed to show_stack and we can't use static log levels.

Eugene, there are a few changes I did to the original patch set:

1. Map the full shadow memory with read and write protection from the beginning. This simplifies a  bit the implementation. I had to use MAP_NORESERVE to have the mapping succeed on Linux.
2. Instead of adding new host ops for KASAN I added options for mmap/munmap/get stack. With some change we should be able to reuse them for MMU support later.
3. I dropped the host kallsym rebuild patch since I found a way to get good .text symbols which should get us good stack traces. .data symbols are still broken
4. Instead of adding stub MMU APIs in LKL I patched KASAN to work for non-MMU systems. I think this is the correct fix and we should be able to upstream these changes easily.

**Comment by tavip** [comment]
@rodionov @thehajime did you get a chance to look at this?

**Comment by thehajime** [comment]
Let me spend sometime to review the patch.  Thanks !

File: arch/lkl/kernel/vmlinux.lds.S [hunk 1]
```diff
@@ -8,28 +8,47 @@ OUTPUT_FORMAT(CONFIG_OUTPUT_FORMAT)
...
+ * from the other adjacent symbols or sections in this file which makes them
+ * bogus.
+ *
+ * Fortunatelly, in most cases their are used for informational purposes only,
+ * e.g. printing the init/code/data size early during boot.
+ */
 SECTIONS
 {
-	__init_begin = .;
-	HEAD_TEXT_SECTION
```
Comment (rodionov) [inline arch/lkl/kernel/vmlinux.lds.S:14]: What about `HEAD_TEXT_SECTION` and `INIT_TEXT_SECTION`? Flagging this just to make sure it's not forgotten.

Comment (tavip) [reply arch/lkl/kernel/vmlinux.lds.S:14]: INIT_TEXT_SECTION expands to 

```
#define INIT_TEXT_SECTION(inittext_align)                               \                                                     
        . = ALIGN(inittext_align);                                      \                                                     
        .init.text : AT(ADDR(.init.text) - LOAD_OFFSET) {               \                                                     
                _sinittext = .;                                         \                                                     
                INIT_TEXT                                               \                                                     
                _einittext = .;                                         \                                                     
        } 
```

and we moved INIT_TEXT and the _sinittext/_einittext assignments directly into the .text section

Comment (rodionov) [reply arch/lkl/kernel/vmlinux.lds.S:14]: Ah, got it. And regarding HEAD_TEXT_SECTION which seems to expand to `KEEP(*(.head.text))` do we need it at all?

Comment (tavip) [reply arch/lkl/kernel/vmlinux.lds.S:14]: HEAD_TEXT_SECTION is used by the arch code for early boot stuff (e.g. decompression). We don't use it. Just to make sure, I double checked it:

```
$ find -name \*.o | xargs objdump -h | grep head.text | wc -l
0
```

**Review by thehajime — COMMENTED** [review]
I run through the whole patchset.  only left minor comments which I wish to confirm.

Plus, with the newly build matrix, we will have a conflict on generated artifact from the ubuntu run and kasan run.  We may need to add a suffix to the filename of the artifact.

https://github.com/lkl/linux/blob/master/.github/workflows/ci.yml#L116

File: tools/lkl/Makefile.autoconf [hunk 2]
```diff
@@ -84,23 +89,60 @@ define nt64_host
...
+  $(call set_kernel_config_h,LKL_CONFIG_64BIT,1)
+  $(call set_kernel_config,64BIT,y)
+endef
+
+define kasan_x86_64
+  $(call set_autoconf_var,KASAN_SHADOW_OFFSET,0x0000200000000000)
+  $(call set_kernel_config,KASAN_SHADOW_OFFSET,0x0000200000000000)
+  $(call set_autoconf_var,KASAN_SHADOW_SIZE,0x0000100000000000)
+  $(call set_kernel_config,KASAN_SHADOW_SIZE,0x0000100000000000)
+  $(if $(filter yes, $(kasan)),$(if $(filter $(1),elf64-x86-64),$(call kasan_x86_64)))
```
[the remark this answers is not in the document]
Comment (tavip) [reply tools/lkl/Makefile.autoconf:126]: No, this is definitely a bug, nice catch! I wonder why it didn’t infinitely recurse…

File: fs/xfs/xfs_buf.c [hunk 3]
```diff
@@ -274,17 +274,24 @@ static void
 xfs_buf_free_pages(
 	struct xfs_buf	*bp)
 {
+#ifdef CONFIG_MMU
```
[the remark this answers is not in the document]
Comment (tavip) [reply fs/xfs/xfs_buf.c:277]: I think this was one of my merges, maybe 5.4? Sorry about that.

File: arch/lkl/kernel/stacktrace.c [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+		const char *loglvl)
+{
+	unsigned int cnt = 0;
+	unsigned long addr;
+	struct stack_frame *frame = __builtin_frame_address(0), *stack = frame;
+
+	if (task != NULL)
+		return;
+
+	printk("%s Call Trace:\n", loglvl);
```
[the remark this answers is not in the document]
Comment (tavip) [reply arch/lkl/kernel/stacktrace.c:29]: Unfortunately we can’t use pr_info, we need to use the log level passed, so we are forced to use printk. I checked other architecture code and it they are doing the same.

File: arch/lkl/Makefile [hunk 5]
```diff
@@ -1,24 +1,20 @@
...
 endif
 # workaround for #include_next<stdarg.h> errors
 LINUXINCLUDE := -isystem arch/lkl/include/system $(LINUXINCLUDE)
 # workaround for https://gcc.gnu.org/bugzilla/show_bug.cgi?id=52991
 KBUILD_CFLAGS += -mno-ms-bitfields
-KBUILD_HOSTCFLAGS += -Wno-char-subscripts
-KBUILD_HOSTLDFLAGS += -s
-else
-$(error Unrecognized platform: $(OUTPUT_FORMAT))
 endif
```
[the remark this answers is not in the document]
Comment (tavip) [reply arch/lkl/Makefile:18]: CONFIG_OUTPUT_FORMAT will be empty when ‘make defconfig’ is called. But we can add a check for a non empty value that us neither pe nor elf32/64.

The other option would be to make this a choice in Kconfig but I don’t know if it works with env vars. I’ll give it a try.

File: arch/lkl/kernel/stacktrace.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (thehajime) [reply arch/lkl/kernel/stacktrace.c:29]: okay, understand.

File: arch/lkl/include/asm/sched.h [hunk 6]
```diff
@@ -11,6 +12,11 @@ static inline void thread_sched_jb(void)
 		set_current_state(TASK_UNINTERRUPTIBLE);
 		lkl_ops->jmp_buf_set(&current_thread_info()->sched_jb,
 				     schedule);
+		/*
+		 * The prevois call to setjmp/longjmp won't unwind the stack
+		 * and, as a result, shadow memory will remain poisoned.
```
[the remark this answers is not in the document]
Comment (tavip) [reply arch/lkl/include/asm/sched.h:17]: Fixed.

File: arch/lkl/Makefile (hunk 5 above)
[the remark this answers is not in the document]
Comment (tavip) [reply arch/lkl/Makefile:20]: Fixed. I added back the check, I will experiment later with Kconfig choice.

**Review by thehajime — APPROVED** [review]
LGTM; feel free to merge it when you're fine, @tavip.

**Pull request merged**